### PR TITLE
[ENH] fABBA Transform

### PR DESCRIPTION
**Pull request by poopsiclepooding**
#### What does this implement/fix? Explain your changes.

This code implements fABBA transform as found in https://github.com/nla-group/fABBA/tree/master . 
fABBA is a time series symbolic representation transform, it compresses and digitizes the time series 
into series of symbols. 
Have added a python implementation of fABBA along with example usage.

Depends on https://github.com/sktime/sktime/pull/8959 for the unequal length output handling, #8959 should be merged first.

#### Does your contribution introduce a new dependency? If yes, which one?

No new dependency.

#### What should a reviewer concentrate their feedback on?

Any formatting or major code issues. Any errors with particular combination of hyperparameters 

#### Did you add any tests for the change?

 Haven't added any particular test. The example provided illustrates the working of algorithm. If any test suggestions are there I can add them.

#### Any other comments?

Original fABBA implementation was really helpful to write this - https://github.com/nla-group/fABBA/tree/master

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the sktime root directory (not the CONTRIBUTORS.md). Common badges: code - fixing a bug, or adding code logic. doc - writing or improving documentation or docstrings. bug - reporting or diagnosing a bug (get this plus code if you also fixed the bug in the PR).maintenance - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] Optionally, for added estimators: I've added myself and possibly to the maintainers tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in docs/source/api_reference/taskname.rst, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant Examples section.

### CODE REVIEW

File: sktime/transformations/series/fabba.py [hunk 1]
```diff
@@ -0,0 +1,1253 @@
+"""fABBA - A Time Series Symbolic Representation.
+
+Implentation derived from following library
+
+{https://github.com/nla-group/fABBA/tree/master?tab=readme-ov-file}.
+
+"""
+
+# Copyright (c) 2025 poopsiclepooding. All rights reserved.
```
Comment (fkiraly) [inline sktime/transformations/series/fabba.py]: you cannot add this code with copyright if you contribute to `sktime`. If you want a restrictive license, you can release in a separate package. Code with `sktime` is released under BSD-3 or compatible permissive license - you can label it with your name though which will be prominently displayed.

File: sktime/transformations/series/fabba.py [hunk 2]
```diff
@@ -0,0 +1,1253 @@
...
+from scipy.sparse.linalg import svds
+from sklearn.cluster import KMeans, MiniBatchKMeans
+
+from sktime.transformations.base import BaseTransformer
+
+
+class fABBA(BaseTransformer):
+    """fABBA - An efficient symbolic aggregate approximation for temporal data.
+
+    fABBA[1] is a dimensionality reduction technique that works by making
```
Comment (fkiraly) [inline sktime/transformations/series/fabba.py]: I think the syntax for getting the reference display properly is `_[1]` - have a look at how the documentation renders by clicking on the documentation CI step below.

File: sktime/transformations/series/fabba.py [hunk 3]
```diff
@@ -0,0 +1,1253 @@
...
+    as compression, clustering, and classification. It converts time series
+    data to sequence of tuples by adaptive polygonal chain approximation. It
+    then uses clustering methods on these tuples and assigns symbols to the
+    clusters.
+
+    Parameters
+    ----------
+    alphabet_size : int, optional (default=5, greater equal 2)
+        descriptive explanation of parama
+    paramb : string, optional (default='default')
```
Comment (fkiraly) [inline sktime/transformations/series/fabba.py]: these have to be replaced and describe the parameters in the `__init__`

File: sktime/transformations/series/fabba.py (hunk 1 above)
Comment (poopsiclepooding) [reply sktime/transformations/series/fabba.py]: Sure will remove that

File: sktime/transformations/series/fabba.py (hunk 3 above)
Comment (poopsiclepooding) [reply sktime/transformations/series/fabba.py]: Okay

File: sktime/transformations/series/fabba.py (hunk 2 above)
Comment (poopsiclepooding) [reply sktime/transformations/series/fabba.py]: Ah okay

File: sktime/transformations/series/fabba.py [hunk 4]
```diff
@@ -0,0 +1,1253 @@
...
+        series = [res.get() for res in result]
+
+        # Return as a df-list
+        return_data = list()
+        for i in range(len(series)):
+            return_data.append(pd.DataFrame({"ts": series[i]}))
+
+        return return_data
+
+    def _get_fitted_params(self):
```
Comment (fkiraly) [inline sktime/transformations/series/fabba.py]: I would recommend: instead of implementing this function, simply write the parameters to `self.cluster_centers_`, `self.num_clusters_`, etc. Then the default will export them.

File: sktime/transformations/series/fabba.py [hunk 5]
```diff
@@ -0,0 +1,1253 @@
...
+                        "scl": 2,
+                        "max_iter": 1,
+                        "auto_digitize": False,
+                        "alpha": 0.7,
+                        "alphabet_set": 0,
+                        "n_jobs": 1,
+                    },
+                ]
+                return params
+        return cls.get_test_params("default")
```
Comment (fkiraly) [inline sktime/transformations/series/fabba.py]: this is equivalent to simply returning `params` - why not that?

Comment (poopsiclepooding) [reply sktime/transformations/series/fabba.py]: Yes okay

**Comment by poopsiclepooding** [comment]
Will look into code formatting.
The implementation in original branch doesn't work for multiple time series of different lengths so I had changed that. 
Also I wanted option to export symbols as int(labels of clusters) rather than just char so implemented that as well.
I used the original repo as reference for main algorithms.

File: sktime/transformations/series/fabba.py (hunk 4 above)
Comment (poopsiclepooding) [reply sktime/transformations/series/fabba.py]: Made the changes

**Comment by fkiraly** [comment]
> I used the original repo as reference for main algorithms.

I see - we should cite it and the paper then.

**Comment by poopsiclepooding** [comment]
Should I add the repo citation below the paper citation or somewhere else as well?

**Comment by fkiraly** [comment]
> Should I add the repo citation below the paper citation or somewhere else as well?

Once in `References`, and then from the main text, I would say.

**Comment by chenxinye** [comment]
Hi @fkiraly and @poopsiclepooding,

Thanks @poopsiclepooding for implementing it, later I will also update my repo to support length variability for time series. 

I think my paper [1] has mentioned this method can be easily extended to arbtrary length time series: 


I suggest the citations can be:
[1] X. Chen, 2024.  Joint symbolic aggregate approximation of time series
[2] X. Chen and S. Güttel., 2024. fABBA: A Python library for the fast symbolic approximation of time series. Journal of Open Source Software 
[3] X. Chen and S. Güttel. 2023. An Efficient Aggregation Method for the Symbolic Representation of Temporal Data. ACM Trans. Knowl. Discov. Data 
[4] S. Elsworth and S. Güttel. 2023.  ABBA: adaptive Brownian bridge-based symbolic aggregation of time serie

Thanks,
Xinye

**Comment by poopsiclepooding** [comment]
Thanks a lot @chenxinye . I will add those citations

**Comment by fkiraly** [comment]
The failures seem to indicate that a single time series is getting transformed intio multiple. Is this expeted behaviour? If yes, we need to set tags differently.

**Comment by poopsiclepooding** [comment]
Yes that is expected behaviour. If we set partition hyperparameter then time series is broken into parts before apply fABBA and each part's symbolic representation is returned. What should I change the tags to?

**Comment by fkiraly** [comment]
there is currently no valid tag for this behaviour, we will have to look into it on framework level. May I suggest to skip the failing tests for now via `tests:skip_by_namd` and raise an issue that explicitly refers to the issue:

if an mtype is passed that can only represent equal length, and the transform changes it to unequal length, the transform tries to convert it back to the original mtype which fails. We need to have a design decision on this.

**Comment by fkiraly** [comment]
@poopsiclepooding, I have updated the framework to handle transformers that turn series into unequal length, here: https://github.com/sktime/sktime/pull/8959

I will merge https://github.com/sktime/sktime/pull/8959 into this Pr and the tests for the fabba transformer should pass now.

FYI, I changed the name to `FABBA` since classes need to start with capital letters.

File: sktime/transformations/series/fabba.py [hunk 6]
```diff
@@ -0,0 +1,1525 @@
...
+
+    .. [3] X. Chen and S. Güttel.
+    An Efficient Aggregation Method for the Symbolic Representation of Temporal Data.
+    ACM Trans. Knowl. Discov. Data (2023).
+    https://dl.acm.org/doi/abs/10.1145/3532622
+
+    .. [4] S. Elsworth and S. Güttel.
+    ABBA: adaptive Brownian bridge-based symbolic aggregation of time serie.
+    https://arxiv.org/abs/2003.12469
+    """
```
Comment (fkiraly) [inline sktime/transformations/series/fabba.py:152]: not blocking, but a usage example would be nice.

Comment (poopsiclepooding) [reply sktime/transformations/series/fabba.py:152]: Sure will add one

**Comment by fkiraly** [comment]
I think the example is now failing, see the failing tests.

**Comment by poopsiclepooding** [comment]
> I think the example is now failing, see the failing tests.

Yes fixed
